### DISCUSSION TITLE
`Optional[Path] = typer.Argument` returns `TypeError` when passed into `open()`

### DISCUSSION THREAD

**Discussion by EndlessTrax** [fastapi/typer#1315] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# cli.py
@app.command()
def main(
    config_file: Optional[Path] = typer.Argument(
        "config.yaml",  # Default if not provided
        exists=True,
        file_okay=True,
        dir_okay=False,
        writable=False,
        readable=True,
        resolve_path=True,
    )
):

    config_data = parse_yaml_config_file(config_file)

# config.py
import yaml

def parse_yaml_config_file(file) -> dict:
    with open(file) as yaml_file:
        return yaml.safe_load(yaml_file)
```

### Description

The above code works as expected in dev, and all unit tests pass. However, after creating a `wheel` and installing into a test `venv` I get the following Traceback:

```
Traceback (most recent call last):
  File "C:\Users\RickyWhite\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\RickyWhite\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\RickyWhite\OneDrive - The Migus Group\Desktop\Test\venv\Scripts\hofund.exe\__main__.py", line 7, in <module>
  File "C:\Users\RickyWhite\OneDrive - The Migus Group\Desktop\Test\venv\lib\site-packages\hofund\cli.py", line 37, in main
    config_data = parse_yaml_config_file(config_file)
  File "C:\Users\RickyWhite\OneDrive - The Migus Group\Desktop\Test\venv\lib\site-packages\hofund\config.py", line 8, in parse_yaml_config_file
    with open(config_file) as yaml_file:
TypeError: expected str, bytes or os.PathLike object, not ArgumentInfo
```

I've check my debugger and with print, and the `config_file` is indeed a `WindowsPath` object.

I've tried converting to a `str`, but get an `OSError: [Errno 22] Invalid argument: '<typer.models.ArgumentInfo object at 0x000001F590930DC0>'` error instead. 

I'm not sure why it seems to be of type `ArgumentInfo` when installed from a `wheel`, but `WindowsPath` object when ran locally. 😕

### Operating System

Windows

### Operating System Details

Windows 11
OS build 22000.438

### Typer Version

0.4.0

### Python Version

3.10

**Comment by Mirandatz** [comment]
Maybe you are using calling main() instead of typer.run(main)?

**Comment by gchhablani** [comment]
I am facing a similar issue when using `typer.Option(None)`.
Basically, I am trying to use `file:str=typer.Option(None)` as default value for a file path input and then trying to read it from somewhere. 

However, during the check `if file is None`, I get a `False` as output.

A snippet for the explaining the basic issue is there:
```python
>>> import typer
>>> var =  typer.Option(None)
>>> var
<typer.models.OptionInfo object at 0x7fc09002d880>
>>> var is None
False
```

The expectation is that `var is None` should return `True`.

This causes the error that I get:
```python
AttributeError: 'OptionInfo' object has no attribute 'open'
``` 
as I am trying to read the file.

**Comment by johnthagen** [comment, marked as answer]
@Mirandatz nailed it ❤️.

In my case I had a entry point defined with Poetry:

```toml
[tool.poetry.scripts]
fact = "fact.cli:main"
```

But I was pointing that to:

```py
def main(n: int = Argument(..., min=0, help="The input n of fact(n)")) -> None:
```

But that wasn't being wrapped in `typer.run(main)`.

I fixed this by pointing the console script at:

```py
def entry_point() -> None:
    typer.run(main)
```
